src/ctia/stores/es/crud.clj

```diff
@@ -37,47 +47,117 @@
            (map #(dissoc % :_id :_index :_type))
            (map coerce!)))))
 
+(s/defn owner? :- s/Bool
+  [doc ident]
+  (boolean
+   (and
+    (= (:owner doc)
+       (:login ident))
+    (clojure.set/subset?
+     (set (:groups ident))
+     (set (:groups doc))))))
+
+(s/defn allow-write? :- s/Bool
+  [current-doc ident]
+  (boolean
+   (cond
+     ;; Document Owner
+     (owner? current-doc ident) true
+
+     ;;or if user is listed in authorized_users field
+     (and (not (nil? (:login ident)))
+          (not (empty? (:authorized_users current-doc)))
+          (clojure.set/subset? #{(:login ident)}
+                               (set (:authorized_users current-doc)))) true
+
+     ;;or if one of the users groups are in the authorized_groups field
+     (and (not (empty? (:authorized_groups current-doc)))
+          (not (empty? (:groups ident)))
+          (clojure.set/subset?
+           (set (:authorized_groups current-doc))
+           (set (:groups ident)))) true
+
+     ;; CTIM models with TLP amber that is owned by org BAR
+     (and (= (:tlp current-doc) "amber")
+          (not (empty? (:groups ident)))
+          (not (empty? (:groups current-doc)))
+          (clojure.set/subset?
+           (set (:groups ident))
+           (set (:groups current-doc)))) true
+
+     ;;CTIM models with TLP red that is owned by user FOO
+     (and (= (:tlp current-doc) "red")
+          (= (:owner current-doc)
+             (:login ident))) true)))
+
+(s/defn allow-read? :- s/Bool
```

Comment:
Could you add tests for `allow-write?`, and  `allow-read?` to check all use cases?